packages/plugin-legacy/src/index.ts

```diff
@@ -258,34 +258,34 @@ function viteLegacyPlugin(options: Options = {}): Plugin[] {
       }
 
       // legacy bundle
-      if (legacyPolyfills.size) {
+      if (options.polyfills !== false) {
         // check if the target needs Promise polyfill because SystemJS relies on it
         // https://github.com/systemjs/systemjs#ie11-support
         await detectPolyfills(
           `Promise.resolve(); Promise.all();`,
           targets,
           legacyPolyfills,
         )
+      }
 
-        isDebug &&
-          console.log(
-            `[@vitejs/plugin-legacy] legacy polyfills:`,
-            legacyPolyfills,
-          )
-
-        await buildPolyfillChunk(
-          config.mode,
+      isDebug &&
+        console.log(
+          `[@vitejs/plugin-legacy] legacy polyfills:`,
           legacyPolyfills,
-          bundle,
-          facadeToLegacyPolyfillMap,
-          // force using terser for legacy polyfill minification, since esbuild
-          // isn't legacy-safe
-          config.build,
-          'iife',
-          opts,
-          options.externalSystemJS,
         )
-      }
+
+      await buildPolyfillChunk(
+        config.mode,
+        legacyPolyfills,
+        bundle,
+        facadeToLegacyPolyfillMap,
+        // force using terser for legacy polyfill minification, since esbuild
+        // isn't legacy-safe
+        config.build,
+        'iife',
+        opts,
+        options.externalSystemJS,
+      )
```

Comment:
To handle one more edge case, should we skip building the polyfill chunk there's no `legacyPolyfills` and `options.externalSystemJS` is `true`?

So we wrap this debug log and `buildPolyfillChunk` call with `if (legacyPolyfills.size || !options.externalSystemJS)`

Reply:
good point, I've updated the code and added a test for `renderModernChunks=false` + `polyfills=false` + `externalSystemJS=true` which no longer produces an empty polyfill chunk.